---
title: "Crohn's disease mRNA-seq"
author: "Michael Love"
output: html_document
---

We start by reading in the sample information, which we call `coldata` because it provides annotation for the *columns* of a count matrix (in genomics, typically the data is transposed to features x samples).

```{r}
library(here)
library(readr)
coldata <- read_delim(here("data","SraRunTable.txt"), delim="\t")
files <- here("data","quants",coldata$Run_s,"quant.sf.gz")
names(files) <- coldata$Run_s
all(file.exists(files))
```

```{r}
suppressPackageStartupMessages(library(GenomicFeatures))
# need to download this file:
if (!file.exists("gencode.v27.sqlite")) {
  # download from web browser:
  # ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_27/gencode.v27.annotation.gtf.gz
  # backup link: 
  # https://www.dropbox.com/s/e4769qm67dzu85k/gencode.v27.annotation.gtf.gz?dl=0
  # takes 100 seconds to make TxDb
  # warnings are OK
  gtf <- makeTxDbFromGFF("gencode.v27.annotation.gtf.gz")
  saveDb(gtf, file="gencode.v27.sqlite")
} else {
  gtf <- loadDb("gencode.v27.sqlite")
}
```

```{r}
columns(gtf)
tx2gene <- select(gtf, keys(gtf, "TXNAME"), "GENEID", "TXNAME")
# backup link:
# https://www.dropbox.com/s/6syciq5uo38qid5/tx2gene.rda?dl=0
#save(tx2gene, file="tx2gene.rda")
```

```{r}
library(tximport)
txi <- tximport(files, type="salmon", tx2gene=tx2gene, dropInfReps=TRUE)
```

```{r}
# tx2gene is always necessary...
coldata$disease_stage_s
coldata$condition <- coldata$disease_stage_s
coldata$condition[is.na(coldata$condition)] <- "normal"
coldata$condition <- sub("/.*","",coldata$condition)
lvls <- c("normal","B1","B2","B3")
coldata$condition <- factor(coldata$condition, lvls)
```


suppressPackageStartupMessages(library(DESeq2))
dds <- DESeqDataSetFromTximport(txi, coldata, ~condition)

# remove failing samples
table(dds$condition)
qcFail <- c("SRR1813874","SRR1813893")
qcFail %in% colnames(dds)
dds <- dds[,!colnames(dds) %in% qcFail]
table(dds$condition)

vsd <- vst(dds)
plotPCA(vsd)
plotPCA(vsd, "Sex_s")
plotPCA(vsd, "age_s")
plotPCA(vsd, "MBases_l")

# add new metadata
gsm <- read_delim(here("data","GSM_table.tsv"),delim="\t",col_names=c("GSM","title"))
samp <- read_delim(here("data","sample_table.tsv"),delim="\t")
gsm$sample <- sub(".*\\[RNA-Seq, (.*?)_.*\\]","\\1",gsm$title)
samp$GSM <- gsm$GSM[match(samp$sample, gsm$sample)]
for (column in c("tissue", "montreal", "batch","sex")) {
  colData(dds)[[column]] <- samp[[column]][match(dds$Sample_Name_s, samp$GSM)]
}
dds$montreal <- factor(dds$montreal, lvls)
dds <- dds[,dds$tissue == "colon"]
table(dds$condition, dds$montreal)
table(dds$Sex_s, dds$sex)

vsd <- vst(dds)
plotPCA(vsd) # old condition
plotPCA(vsd, "montreal")
plotPCA(vsd, "batch")
plotPCA(vsd, "sex")

design(dds) <- ~montreal

dds <- estimateSizeFactors(dds)
keep <- rowSums(counts(dds, normalized=TRUE) >= 10) >= 5
table(keep)
dds <- dds[keep,]

# takes <1 min
system.time({ dds <- DESeq(dds, test="LRT", reduced=~1, minReplicatesForReplace=Inf) })
res <- results(dds, alpha=.05)
summary(res)

resultsNames(dds)
system.time({
  # takes 30 s per call
  lfcB3 <- lfcShrink(dds, coef=4, type="normal")
})
lfcB2 <- lfcShrink(dds, coef=3, type="normal")
lfcB1 <- lfcShrink(dds, coef=2, type="normal")

plotMA(lfcB3, xlim=c(1,1e6), ylim=c(-2,2), cex=1)
plot(lfcB3$log2FoldChange, lfcB2$log2FoldChange)
abline(v=0, h=0, col="red")
plot(lfcB3$log2FoldChange, lfcB1$log2FoldChange)
abline(v=0, h=0, col="red")

par(mfrow=c(3,3))
for (i in 1:9) 
  plotCounts(dds, gene=order(res$pvalue)[i], transform=FALSE)
par(mfrow=c(1,1))

rownames(res)[order(res$pvalue)[1:9]]

library(org.Hs.eg.db)
unname(
  mapIds(org.Hs.eg.db,
       sub("\\..*","",rownames(res)[order(res$pvalue)[1:50]]),
       "SYMBOL",
       "ENSEMBL")
)

library(org.Hs.eg.db)
unname(
  mapIds(org.Hs.eg.db,
         sub("\\..*","",rownames(res)[order(res$pvalue)[1:50]]),
         "GENENAME",
         "ENSEMBL")
)

library(goseq)
padj <- ifelse(is.na(res$padj), 1, res$padj)
genes <- as.integer(padj < .05)
strp <- function(x) sub("(.*)\\..*","\\1",x)
names(genes) <- strp(rownames(res))
table(genes)
table(duplicated(names(genes)))
table(genes[duplicated(names(genes))])
genes <- genes[!duplicated(names(genes))]

supportedOrganisms()[supportedOrganisms()$Genome=="hg38",]

# we need length

# easy way (tximport gives us gene lengths)
lengthData <- rowMeans(assays(dds)[["avgTxLength"]])
names(lengthData) <- strp(names(lengthData))
lengthData <- lengthData[names(genes)]

# what if we didn't use tximport?
ebg <- exonsBy(gtf, by="gene")
rangePlotter <- function(x) {
  plot(c(start(x),end(x)),rep(seq_along(x),2),cex=.5,xlab="bp",ylab="")
  segments(start(x),seq_along(x),end(x),seq_along(x),lwd=3)
}
rangePlotter(ebg[[1]])
ebg.red <- reduce(ebg)
rangePlotter(ebg.red[[1]])
lengthData2 <- sum(width(ebg))
names(lengthData2) <- strp(names(lengthData2))
lengthData2 <- lengthData2[names(genes)]

# compare the two
plot(lengthData2, lengthData, log="xy", cex=.1,
     xlab="sum of all exons", ylab="length of expressed transcripts")
abline(0,1,col="red")

pwf <- nullp(genes, genome="hg39", id="ensGene", bias.data=lengthData)
go <- goseq(pwf, "hg38", "ensGene", test.cats="GO:BP")

head(subset(go, numInCat < 200),20)
```

